### PR TITLE
feat(cdk-graph): add threat composer plugin for generating base threat models

### PR DESCRIPTION
**Pull request by cogwirrel**
This adds a new CDK Graph plugin which generates Threat Composer threat models from PDKNag apps.

Threat Composer: https://github.com/awslabs/threat-composer/

The generated threat model provides a starting point to build upon for threat modelling an application.

### CODE REVIEW

**Comment by cogwirrel** [comment]
Hmm the build is failing with:

```
@aws/pipeline: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
```

Looks like a Monday problem 😄 

Edit: Seems ok now, may be reaching task runner memory limits so might need to reduce parallelism or use a larger worker in future.

File: packages/cdk-graph-plugin-threat-composer/src/model-generator/threat-model-generator.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+import { orderBy, uniq } from "lodash";
+import * as BASE_THREAT_MODEL from "./base-model/threat-composer-base-model.tc.json";
+import { ThreatComposerApplicationDetails } from "./types";
+
+type ThreatComposerModel = typeof BASE_THREAT_MODEL;
+type ThreatComposerThreat = ThreatComposerModel["threats"][number];
+type ThreatComposerMitigation = ThreatComposerModel["mitigations"][number];
+
+interface ComplianceCount {
+  compliantResources: Set<string>;
```
Comment (agdimech) [inline packages/cdk-graph-plugin-threat-composer/src/model-generator/threat-model-generator.ts]: should these be marked as `readonly`?

File: packages/cdk-graph-plugin-threat-composer/src/model-generator/threat-model-generator.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+import { ExtendedNagResult, NagResultCompliance } from "@aws/pdk-nag";
+import { orderBy, uniq } from "lodash";
+import * as BASE_THREAT_MODEL from "./base-model/threat-composer-base-model.tc.json";
+import { ThreatComposerApplicationDetails } from "./types";
+
+type ThreatComposerModel = typeof BASE_THREAT_MODEL;
+type ThreatComposerThreat = ThreatComposerModel["threats"][number];
+type ThreatComposerMitigation = ThreatComposerModel["mitigations"][number];
+
+interface ComplianceCount {
```
Comment (agdimech) [inline packages/cdk-graph-plugin-threat-composer/src/model-generator/threat-model-generator.ts]: Interface name doesn't really match the definition. Perhaps rename to `ComplianceFindings`?

File: packages/cdk-graph-plugin-threat-composer/src/plugin.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+export class CdkGraphThreatComposerPlugin implements ICdkGraphPlugin {
+  /**
+   * Fixed ID of the threat-composer plugin
+   */
+  static readonly ID = "threat-composer";
+
+  /**
+   * Curent semantic version of the threat-composer plugin
+   */
+  static readonly VERSION = "0.0.0";
```
Comment (agdimech) [inline packages/cdk-graph-plugin-threat-composer/src/plugin.ts:40]: How will this be maintained?

File: packages/cdk-graph-plugin-threat-composer/src/model-generator/threat-model-generator.ts (hunk 1 above)
Comment (cogwirrel) [reply packages/cdk-graph-plugin-threat-composer/src/model-generator/threat-model-generator.ts]: yep! I'd previously had them as counters but swapped to sets, will update :)

File: packages/cdk-graph-plugin-threat-composer/src/plugin.ts (hunk 3 above)
Comment (cogwirrel) [reply packages/cdk-graph-plugin-threat-composer/src/plugin.ts:40]: I think this is only for cdk graph plugin dependencies - it's a required property so I followed suit with the diagram plugin. From a quick look at the cdk-graph code I couldn't see how dependencies were honoured within the plugin framework though, but I might not have looked hard enough!

**Review by agdimech — APPROVED** [review]

File: packages/cdk-graph-plugin-threat-composer/docs/developer_guides/cdk-graph-plugin-threat-composer/index.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+!!!warning
+    This plugin must be used in tandem with [pdk-nag](../../developer_guides/pdk-nag/index.md), using `PDKNag.app()`
+
+> This plugin currently only supports `async report()` generation following the above example. **Make sure to wrap the cdk app with _async IIFE_.**
+
+### How it Works
+
+This plugin uses the [CDK Nag](https://github.com/cdklabs/cdk-nag) findings from your [pdk-nag](../../developer_guides/pdk-nag/index.md) app to generate a [Threat Composer](https://github.com/awslabs/threat-composer) threat model for you.
```
Comment (dboyd13) [inline packages/cdk-graph-plugin-threat-composer/docs/developer_guides/cdk-graph-plugin-threat-composer/index.md]: "generates a starter/base threat model for you."

File: packages/cdk-graph-plugin-threat-composer/docs/developer_guides/cdk-graph-plugin-threat-composer/index.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+Upon configuring the plugin and synthesizing CDK, you'll find a `cdk.out/cdkgraph/threat-composer.tc.json` file in your project. You can import this into the [Threat Composer Website](https://awslabs.github.io/threat-composer/) to view and edit your threat model.
```
Comment (dboyd13) [inline packages/cdk-graph-plugin-threat-composer/docs/developer_guides/cdk-graph-plugin-threat-composer/index.md]: "into the [Threat Composer](https://awslabs.github.io/threat-composer/) tool"

File: packages/cdk-graph-plugin-threat-composer/docs/developer_guides/cdk-graph-plugin-threat-composer/index.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+!!!note
+    The generated threat model is not complete, however it provides a good starting point for you to consider possible threats to your application and how they are mitigated.
+
+    After editing your threat model, it's recommended that you manage it as part of your codebase.
```
Comment (dboyd13) [inline packages/cdk-graph-plugin-threat-composer/docs/developer_guides/cdk-graph-plugin-threat-composer/index.md]: ... by checking into version control.

File: packages/cdk-graph-plugin-threat-composer/src/model-generator/base-model/README.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+## Modifying the Base Threat Model
+
+This folder contains a base threat model (`.tc.json` file) with example threats mapped to CDK Nag rules, and the plugin filters to only the applicable threats to the CDK project.
+
+The easiest way to modify the threat model is to import the json file in the [Threat Composer Website](https://awslabs.github.io/threat-composer/), modify it using the UI, and then export it as json and replace the contents in this repository.
```
Comment (dboyd13) [inline packages/cdk-graph-plugin-threat-composer/src/model-generator/base-model/README.md]: "is to import the json file into the [Threat Composer](https://awslabs.github.io/threat-composer/) tool"

**Review by dboyd13 — COMMENTED** [review]
Couple of documentation suggestions

**Comment by dboyd13** [comment]
👍

**Pull request merged**